pkg/controller/issuers/ca/ca.go

```diff
@@ -0,0 +1,147 @@
+		issuerOptions: ctx.IssuerOptions,
+		recorder:      ctx.Recorder,
+	}
+}
+
+func init() {
+	// create issuer controller for ca
+	controllerpkg.Register(IssuerControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, IssuerControllerName).
+			For(issuers.New(IssuerControllerName, cmapi.IssuerKind, New(ctx))).
+			Complete()
+	})
+
+	// create cluster issuer controller for ca
+	controllerpkg.Register(ClusterIssuerControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, ClusterIssuerControllerName).
+			For(issuers.New(ClusterIssuerControllerName, cmapi.ClusterIssuerKind, New(ctx))).
+			Complete()
+	})
+}
+
+func (c *CA) Setup(ctx context.Context, issuer cmapi.GenericIssuer) error {
+	// Namespace in which to read resources related to this Issuer from.
+	// For Issuers, this will be the namespace of the Issuer.
+	// For ClusterIssuers, this will be the cluster resource namespace.
+	resourceNamespace := c.issuerOptions.ResourceNamespace(issuer)
+
+	log := logf.FromContext(ctx, "setup").WithName(issuer.GetName())
+
+	cert, err := utilkube.SecretTLSCert(ctx, c.secretsLister, resourceNamespace, issuer.GetSpec().CA.SecretName)
+	if err != nil {
+		log.Error(err, "error getting signing CA TLS certificate")
+		s := messageErrorGetKeyPair + err.Error()
+		c.recorder.Event(issuer, corev1.EventTypeWarning, errorGetKeyPair, s)
+		apiutil.SetIssuerCondition(issuer, cmapi.IssuerConditionReady, cmmeta.ConditionFalse, errorGetKeyPair, s)
+		return err
+	}
+
+	_, err = utilkube.SecretTLSKey(ctx, c.secretsLister, resourceNamespace, issuer.GetSpec().CA.SecretName)
+	if err != nil {
+		log.Error(err, "error getting signing CA private key")
+		s := messageErrorGetKeyPair + err.Error()
+		c.recorder.Event(issuer, corev1.EventTypeWarning, errorGetKeyPair, s)
+		apiutil.SetIssuerCondition(issuer, cmapi.IssuerConditionReady, cmmeta.ConditionFalse, errorGetKeyPair, s)
+		return err
+	}
+
+	log = logf.WithRelatedResourceName(log, issuer.GetSpec().CA.SecretName, resourceNamespace, "Secret")
+	if !cert.IsCA {
+		s := messageErrorGetKeyPair + "certificate is not a CA"
+		log.Error(nil, "signing certificate is not a CA")
+		c.recorder.Event(issuer, corev1.EventTypeWarning, errorInvalidKeyPair, s)
+		apiutil.SetIssuerCondition(issuer, cmapi.IssuerConditionReady, cmmeta.ConditionFalse, errorInvalidKeyPair, s)
+		// Don't return an error here as there is nothing more we can do
+		return nil
+	}
+
+	log.Info("signing CA verified")
+	c.recorder.Event(issuer, corev1.EventTypeNormal, successKeyPairVerified, messageKeyPairVerified)
+	apiutil.SetIssuerCondition(issuer, cmapi.IssuerConditionReady, cmmeta.ConditionTrue, successKeyPairVerified, messageKeyPairVerified)
+
+	return nil
+}
+
+func (c *CA) Implements(issuer cmapi.GenericIssuer) bool {
+	return issuer.GetSpec().CA != nil
```

Comment:
How about if   `Implements` returned an `interface{}` and then each issuer could return e.g. `&VenafiIssuerSpec{}` which would feed in to a type switch higher up....a bit hand wavy, sorry.

Reply:
How would this really work? You can't really do this type switch with the field being different for each issuer.